### PR TITLE
UAR-347: Review navigation controls on all update pages

### PR DESCRIPTION
**Pull request by bbell-companieshouse**
### JIRA link
https://companieshouse.atlassian.net/browse/UAR-347

### Change description
Review navigation controls on all update pages and make appropriate code changes

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/routes/index.ts [hunk 1]
```diff
@@ -329,15 +329,6 @@ router.route(config.OVERSEAS_ENTITY_PRESENTER_URL)
   .get(overseasEntityPresenter.get)
   .post(...validator.presenter, checkValidations, overseasEntityPresenter.post);
 
-router.route(config.UPDATE_CHECK_YOUR_ANSWERS_URL)
```
Comment (AndreKramer7654) [inline src/routes/index.ts:332]: note for other reviewers: this is a duplicate route

**Review by AndreKramer7654 — APPROVED** [review]

File: src/middleware/navigation/check.condition.ts [hunk 2]
```diff
@@ -88,3 +88,11 @@ export const checkEntityUpdateDetailsEntered = (appData: ApplicationData): boole
 export const checkWhoIsFilingEntered = (appData: ApplicationData): boolean => {
   return checkHasAppData(appData) && (appData[WhoIsRegisteringKey] || "").length !== 0;
 };
+
```
Comment (AndreKramer7654) [inline src/middleware/navigation/check.condition.ts:91]: checkBoDetailsdEntered() || checkMOsDetailsEntered()

File: src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { Request, Response, NextFunction } from 'express';
+
+import { logger } from '../../../utils/logger';
+import { UPDATE_BENEFICIAL_OWNER_TYPE_URL } from '../../../config';
+import { getApplicationData } from "../../../utils/application.data";
+import { checkBOsOrMOsDetailsEnteredUpdate, NavigationErrorMessage } from '../check.condition';
+
+export const hasBOsOrMOsUpdate = (req: Request, res: Response, next: NextFunction): void => {
+  try {
+    if ( !checkBOsOrMOsDetailsEnteredUpdate(getApplicationData(req.session)) ) {
```
Comment (AndreKramer7654) [inline src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts:10]: have you "make lint" on this?

File: src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      logger.infoRequest(req, NavigationErrorMessage);
+      return res.redirect(UPDATE_BENEFICIAL_OWNER_TYPE_URL);
```
Comment (stephanie-ch) [inline src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts:12]: Did this redirect location come from product? I've a faint memory of product mentioning navigation check failures should take us to the beginning of the journey. Worth double checking :)

File: src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts (hunk 3 above)
Comment (bbell-companieshouse) [reply src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts:10]: Passes the lint check okay, although whitespace has been removed

File: src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts (hunk 4 above)
Comment (bbell-companieshouse) [reply src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts:12]: Following discussion with Amelia Ward it was agreed to send user back to BO/MO type page. In register journey on the check-your-answers navigation, there is a check for BOs and MOs which sends the user back to the sold land filter page- this page does not exist in update so we agreed at the time it should be changed to the BO/MO page. However upon reflection this should probably be the secure update filter page to fit in line with the other validation checks. I will  confirm this with product

Comment (bbell-companieshouse) [reply src/middleware/navigation/update/has.beneficial.owners.or.managing.officers.update.middleware.ts]: This has been changed to send the user back to the start of the journey (secure update filter)

File: src/middleware/navigation/check.condition.ts [hunk 5]
```diff
@@ -21,6 +21,17 @@ const checkHasAppData = (appData: ApplicationData): boolean => {
   return appData && Object.keys(appData).length !== 0;
 };
 
+const checkBOsDetailsEntered = (appData: ApplicationData): boolean => {
```
Comment (AndreKramer7654) [inline src/middleware/navigation/check.condition.ts:24]: I meant these are already defined elsewhere.

Comment (AndreKramer7654) [reply src/middleware/navigation/check.condition.ts:24]: in application.data.ts

export const checkBOsDetailsEntered = (appData: ApplicationData): boolean => {
  return Boolean( appData[BeneficialOwnerIndividualKey]?.length || appData[BeneficialOwnerOtherKey]?.length || appData[BeneficialOwnerGovKey]?.length );
};

export const checkMOsDetailsEntered = (appData: ApplicationData): boolean => {
  return Boolean( appData[ManagingOfficerKey]?.length || appData[ManagingOfficerCorporateKey]?.length ) ;
};

Comment (AndreKramer7654) [reply src/middleware/navigation/check.condition.ts:24]: Should have made that clear, sorry

File: src/middleware/navigation/check.condition.ts [hunk 6]
```diff
@@ -5,15 +5,12 @@ import { HasSoldLandKey, IsSecureRegisterKey, EntityNameKey, EntityNumberKey } f
...
-import { BeneficialOwnerGovKey } from '../../model/beneficial.owner.gov.model';
-import { BeneficialOwnerIndividualKey } from '../../model/beneficial.owner.individual.model';
-import { BeneficialOwnerOtherKey } from '../../model/beneficial.owner.other.model';
-import { ManagingOfficerCorporateKey } from '../../model/managing.officer.corporate.model';
-import { ManagingOfficerKey } from '../../model/managing.officer.model';
 import { WhoIsRegisteringKey } from '../../model/who.is.making.filing.model';
 import { OverseasEntityDueDiligenceKey } from '../../model/overseas.entity.due.diligence.model';
 import { DueDiligenceKey } from '../../model/due.diligence.model';
 import { UpdateKey } from '../../model/update.type.model';
+import { checkBOsDetailsEntered } from '../../utils/application.data';
```
Comment (AndreKramer7654) [inline src/middleware/navigation/check.condition.ts:12]: Do both in one line
import { checkBOsDetailsEntered, checkMOsDetailsEntered } from '../../utils/application.data';
Surprised lint did not complain

**Review by AndreKramer7654 — APPROVED** [review]

**Pull request merged**